### PR TITLE
fix: saving files to downloads folder and empty asset names on older Androids [AR-2985] [AR-2984]

### PR DESCRIPTION
**Pull request by saleniuk**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-2985" title="AR-2985" target="_blank">BugAR-2985</a>  Can't save files on android 9
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [x] contains a reference JIRA issue number like `SQPIT-764`
    - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

- Crash when saving files. [AR-2984](https://wearezeta.atlassian.net/browse/AR-2984)
- Not possible to save files on Android 9. [AR-2985](https://wearezeta.atlassian.net/browse/AR-2985)

### Causes (Optional)

- Crash is probably because of the empty asset name, it would explain why there's an error saying that it's a directory and not a file. Asset names actually are only optional so it can happen.
- For Android older than Q (29), we get downloads directory as `getExternalFilesDir(DIRECTORY_DOWNLOADS)` but this actually creates "downloads" directory inside the private "files" directory for our app instead of returning the public "downloads" folder.
Another thing it that on newer Androids, system automatically handles the case of duplicates, so if you want to add to "downloads" a file that already exists, it will rename it so you can have both, but on older versions we need to do it ourselves otherwise the file will be overwritten.

### Solutions

- Provide some generic file name "attachment" when the asset name is empty.
- Use `Environment.getExternalStoragePublicDirectory` instead to get the public "downloads" directory.
- Ask for `WRITE_EXTERNAL_STORAGE` permission.
- Check and find an unique name if the file already exists in the given directory to not overwrite - add copy counter to the file name like this: `filename (2).jpg`.

### Dependencies (Optional)

Needs releases with:

- https://github.com/wireapp/kalium/pull/1410

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

Try to save attachment to the "downloads" directory on Android 9.

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

[AR-2984]: https://wearezeta.atlassian.net/browse/AR-2984?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AR-2985]: https://wearezeta.atlassian.net/browse/AR-2985?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-1386/1/) **failed**.

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1386/2/) **failed**.

**Comment by AndroidBob** [comment]
Build [3](https://10.10.124.134/job/wire-android-reloaded/job/PR-1386/3/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5585197)(5585197)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv0.0.1(5585197)(5585197)-staging-compat.apk)

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-1386/4/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5618801)(5618801)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv0.0.1(5618801)(5618801)-staging-compat.apk)

**Review by gongracr — APPROVED** [review]
Looking great 💪 🍄 ⭐  Let's try to follow up on analytics whether these fixes are actually tackling the issue or not over the next weeks after release

File: app/src/main/kotlin/com/wire/android/util/FileUtil.kt [hunk 1]
```diff
@@ -93,14 +97,18 @@ private fun Context.saveFileDataToDownloadsFolder(assetName: String, downloadedD
     val resolver = contentResolver
     return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
         val contentValues = ContentValues().apply {
-            put(MediaStore.MediaColumns.DISPLAY_NAME, assetName)
+            // creates a copy with a different name if such file already exists
+            put(MediaStore.MediaColumns.DISPLAY_NAME, assetName.ifEmpty { ATTACHMENT_FILENAME })
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/util/FileUtil.kt:101]: Here this is jsut applying a non-empty asset name, not really making a copy. Is this what you intended or the comment is wrong?

Comment (saleniuk) [reply app/src/main/kotlin/com/wire/android/util/FileUtil.kt:101]: I meant that the `ContentResolver` creates a copy, but actually it just uses the different name in that case, I will write this comment more precisely. 😅

**Comment by AndroidBob** [comment]
Build [5](https://10.10.124.134/job/wire-android-reloaded/job/PR-1386/5/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.0.0(5630090)(5630090)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.0.0(5630090)(5630090)-staging-compat.apk)

**Comment by AndroidBob** [comment]
Build [6](https://10.10.124.134/job/wire-android-reloaded/job/PR-1386/6/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.0.0(5630324)(5630324)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.0.0(5630324)(5630324)-staging-compat.apk)

**Pull request merged**